src/component.jl

```diff
@@ -184,29 +214,149 @@ Base.sortperm(c::SharedComponent) = sortperm(c.data)
+                    end
+                end
+            else
+                fn_to_DT[fn] = DT
+                fnq = QuoteNode(fn)
+                ex = Expr(:elseif, :(f === $fnq), :(return getfield(e[$DT], $fnq)::$ft), ex)
+            end
+        end
+    end
+    ex.head = :if
+    return ex
+end
+
+@generated function Base.setproperty!(e::EntityState{TT}, f::Symbol, val) where {TT}
+    fn_to_DT = Dict{Symbol, DataType}()
+    ex = :(getfield(e, f))
+    ex = Expr(:elseif, :(f === :id), :(return setfield!(getfield(e, :e), :id)::Int, val), ex)
+    for PDT in TT.parameters
+        DT = eltype(PDT)
+        for (fn, ft) in zip(fieldnames(DT), fieldtypes(DT))
+            if haskey(fn_to_DT, fn)
+                fnq = QuoteNode(fn)
+                DT_ = fn_to_DT[fn]
+                ft_ = fieldtype(DT_, fn)
+                ex = MacroTools.postwalk(ex) do x
+                    if @capture(x, setfield!(t, $fnq, val))
+                        return quote
+                            throw(error("Field $f found in multiple components in $e.\nPlease use entity_state[$($DT)].$f instead."))
+                        end
+                    else
+                        return x
+                    end
+                end
+            else
+                fn_to_DT[fn] = DT
+                fnq = QuoteNode(fn)
+                tex = quote
+                    ptr = valid_pointer(e, $DT)
+                    t = Base.unsafe_pointer_to_objref(ptr)
+                    setfield!(t, $fnq, val)
+                end
+                    
+                ex = Expr(:elseif, :(f === $fnq), tex, ex)
+            end
+        end
+    end
+    ex.head = :if
+    return ex
+end
+
+@generated function Base.pointer(e::EntityState{TT}, ::Type{T}) where {TT<:Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, TT.parameters)
+    return :(getfield(e, :pointers)[$id])
+end
+
+#TODO: These boundchecks etc are to ensure in the case of || in @entities_in nobody is trying to load some
+#      invalid memory. Do we really need ||?
+#      diff in benchmarks is 47.8 vs 48.2 mus
+#      Solution, have Ptr and MaybePtr so that the boundscheck can be done at compile time,
+#      or have a different iterator for || vs &&
+
+@inline function valid_pointer(e::EntityState, ::Type{T}) where {T<:ComponentData}
+    ptr = pointer(e, T)
+    if ptr === Ptr{T}()
+        throw(error("$e does not have ComponentData $T"))
+    end
+    return ptr
+end
+
+Base.@propagate_inbounds @inline pointers(x::Tuple{}, e::Entity) = ()
+Base.@propagate_inbounds @inline pointers(x::Tuple, e::Entity) = (pointer(first(x), e), pointers(Base.tail(x), e)...)
+
+Base.getindex(e::EntityState, ::Type{T}) where {T<:ComponentData} =
+        T.mutable ? Base.unsafe_pointer_to_objref(valid_pointer(e, T))::T : unsafe_load(valid_pointer(e, T))::T
+
+Base.setindex!(e::EntityState, x::T, ::Type{T}) where {T<:ComponentData} =
+    unsafe_store!(valid_pointer(e, T), x)
```

Comment:
Unfortunately I think this is invalid if `x` is `mutable` because it can break invariants relied on by the generational GC.  In particular, it could result in an object of the old generation pointing to a new generation object, without the GC being aware of it.

Reply:
I guess you mean when you are assigning a completely new Mutable to that pointer? Yea I'm not that familiar sadly with exactly how the GC works, I should really read up on it.